src/PhpDoc/TypeNodeResolver.php

```diff
@@ -505,6 +506,12 @@ private function resolveGenericTypeNode(GenericTypeNode $typeNode, NameScope $na
 
 				return IntegerRangeType::fromInterval($min, $max);
 			}
+		} elseif ($mainTypeName === 'key-of') {
+			if (count($genericTypes) == 1 && $genericTypes[0] instanceof ConstantArrayType) { // key-of<ValueType>
```

Comment:
The `instanceof` shouldn't be here, just the `count` part.

Reply:
According to your main comment, I understand that this check should be part of a custom rule, right?

Reply:
No need for a custom rule, types that don't have a sense of "iterable key" will just return ErrorType and that will get picked up by existing rules.